### PR TITLE
Optimize the API

### PR DESCRIPTION
**Pull request by PythonFZ**
This PR removes the dependency on `self.dvc` and instead introduces `pytrack.out` and other to do this for you.

```py
from pytrack import PyTrack, PyTrackProject, DVC

@PyTrack(nb_name="Test01.ipynb")
class Stage:
    def __init__(self):
        self.param = DVC.parameter()
        self.result1 = DVC.result()
        self.result2 = DVC.result()
        self.model= DVC.deps(["model.yml", "model.pckl"])
        self.file = DVC.deps()
        self.out= DVC.out("Output1.yaml")
        self.metric = DVC.out()
    
    def __call__(self, param1, param2, file):
        self.param1 = {"param1": param1, "param2": param2}
        self.file= file
        self.metric = "metric.yaml"
        
        
    def run(self):
        self.result1 = self.param1 * self.param2
        self.result2 = "This is a second result"
        self.out.write_text(self.model[0].read_text())
        print(self.model)
        print(self.metric)
```

Notes
---------
EDIT
A lot has been rewriten in terms of descriptors.
This leads to a lot of required testing, because they can get very difficult to debug, but they allow more readable PyTrack code later and I think it makes sense to use them.

Still in testing are
- ~~lists of all kind with `DVC.<placeholder>`#~~
- ~~PyTrack.results as dependendency currently does not work, maybe make it `DVC.deps(PyTrack)` possible or keep `DVC.deps(PyTrack.json_file)` ? First one would not require a "hidden" attribute only set by the decorator.~~

EDIT END

- ~~jupyter mode does not support comments that start the first line with #~~
- remove repeating code
- try to make code more readable
- currently only supports params, results, dependecies and outs!
- ~~access to `self.outs` is not possible, they have to be overwritten after the `__init__`~~
- ~~Tests still fail~~
- missing dvc options, not all are implemented
- ~~consider having `from pytrack import dvc` and then `dvc.outs()`~~
- ~~`__call__` should only be called once per instance, reset or raise warning!~~

Prior Code:

```py
from pytrack import PyTrack, DVCParams

@PyTrack(nb_name="Test01.ipynb")
class Stage:
    def __init__(self):
        self.dvc = DVCParams(outs=["Model", "metric"])
        self.param1 = parameter()
        self.param2 = parameter()
        self.result1 = result()
        self.result2 = result()

    
    def __call__(self, param1, param2, deps):
        self.param1 = param1
        self.param2 = param2
        self.dvc.deps  = deps        
        
    def run(self):
        self.result1 = self.param1 * self.param2
        self.result2 = "This is a second result"
        print(self.dvc.outs[0])
        print(self.dvc.outs[1])
```

### CODE REVIEW

File: CI/integration_tests/test_graph01.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
+"""
```
Comment (SamTov) [inline CI/integration_tests/test_graph01.py:1]: Can you put an underscore before the 01 i.e. test_graph_01

File: pytrack/core/data_classes.py [hunk 2]
```diff
@@ -23,8 +23,7 @@
 class DVCParams:
     # pytrack Parameter
```
Comment (SamTov) [inline pytrack/core/data_classes.py:24]: doc string?

File: pytrack/core/parameter.py [hunk 3]
```diff
@@ -9,47 +9,180 @@
...
 """
 
+from pytrack.utils import is_jsonable
+from pytrack.core.data_classes import DVCParams
+from pathlib import Path
+from typing import Union
+
+
+class ParameterHandler:
+    def __init__(self):
```
Comment (SamTov) [inline pytrack/core/parameter.py:19]: class doc string.

File: pytrack/core/py_track.py [hunk 4]
```diff
@@ -29,19 +27,9 @@
 class PyTrackParent:
```
Comment (SamTov) [inline pytrack/core/py_track.py:27]: class doc string

File: pytrack/project/py_track_project.py [hunk 5]
```diff
@@ -12,13 +12,16 @@
 
 from pytrack.interface import DVCInterface
 import subprocess
+from datetime import datetime
 
 log = logging.getLogger(__file__)
 
 
 class PyTrackProject(DVCInterface):
```
Comment (SamTov) [inline pytrack/project/py_track_project.py:20]: class doc string

File: pytrack/utils/utils.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+This program and the accompanying materials are made available under the terms of the
+Eclipse Public License v2.0 which accompanies this distribution, and is available at
+https://www.eclipse.org/legal/epl-v20.html
+SPDX-License-Identifier: EPL-2.0
+
+Copyright Contributors to the Zincware Project.
+
+Description:
+"""
+
```
Comment (SamTov) [inline pytrack/utils/utils.py:11]: too much space here?

**Review by SamTov — CHANGES_REQUESTED** [review]
Please address the doc strings.

File: CI/integration_tests/test_graph01.py (hunk 1 above)
Comment (SamTov) [reply CI/integration_tests/test_graph01.py:1]: still stands.

**Review by SamTov — COMMENTED** [review]
Comments from previous review and not resolved.

**Review by SamTov — COMMENTED** [review]
Love the new example.

File: CI/integration_tests/test_graph01.py (hunk 1 above)
Comment (PythonFZ) [reply CI/integration_tests/test_graph01.py]: solved in 2175398a95c5813644aef5da0d41bc8ce2df9d21

File: pytrack/core/parameter.py (hunk 3 above)
Comment (PythonFZ) [reply pytrack/core/parameter.py]: solved in 2175398a95c5813644aef5da0d41bc8ce2df9d21

File: pytrack/project/py_track_project.py (hunk 5 above)
Comment (PythonFZ) [reply pytrack/project/py_track_project.py:20]: solved in 2175398a95c5813644aef5da0d41bc8ce2df9d21

File: pytrack/utils/utils.py (hunk 6 above)
Comment (PythonFZ) [reply pytrack/utils/utils.py:11]: solved in 2175398a95c5813644aef5da0d41bc8ce2df9d21

File: pytrack/core/data_classes.py (hunk 2 above)
Comment (PythonFZ) [reply pytrack/core/data_classes.py:25]: 2175398a95c5813644aef5da0d41bc8ce2df9d21

File: pytrack/core/py_track.py (hunk 4 above)
Comment (PythonFZ) [reply pytrack/core/py_track.py:26]: solved in 2175398a95c5813644aef5da0d41bc8ce2df9d21

File: pytrack/core/py_track.py [hunk 7]
```diff
@@ -10,57 +10,53 @@
...
-import json
 import subprocess
 import yaml
-from pathlib import Path
-import abc
 
-from .dataclasses import DVCParams, SlurmConfig
-from .parameter import parameter, result
-from pytrack.utils import is_jsonable
+from .data_classes import SlurmConfig
```
Comment (SamTov) [inline pytrack/core/py_track.py:18]: Have we done any slurm testing?

**Review by SamTov — CHANGES_REQUESTED** [review]
The code looks good but I will add some further documentation tonight and merge when it is finished.

File: pytrack/core/py_track.py (hunk 7 above)
Comment (PythonFZ) [reply pytrack/core/py_track.py:18]: Yes, it uses `srun -n n_configs python -f from <module> import <stage>; <stage>.(id_=0).run()`

File: docs/source/installation.rst [hunk 8]
```diff
@@ -1,21 +1,32 @@
...
 
 If you want to install PyTrack from source you can use
 
-.. code-block::
+.. code-block:: bash
 
     git clone https://github.com/zincware/py-track.git
-    pip install .
+    cd py-track
+    pip3 install . --user
```
Comment (PythonFZ) [inline docs/source/installation.rst:16]: On Windows I had issues with the `--user` and I would not recommend using it as a default.

Comment (SamTov) [reply docs/source/installation.rst:16]: Okay

**Review by SamTov — APPROVED** [review]
Still doc strings to be written but otherwise looks good.

**Pull request merged**